lib/Monads/trace/Trace_No_Trace.thy

```diff
@@ -44,34 +44,318 @@ lemma no_trace_is_triple[wp_trip]:
+  "(\<And>m. m \<in> f ` set xs \<Longrightarrow> no_trace m) \<Longrightarrow> no_trace (mapME f xs)"
+  unfolding mapME_def
+  by (fastforce intro: no_trace_sequenceE)
+
+lemma no_trace_mapME_x[no_trace_cond]:
+  "(\<And>m'. m' \<in> f ` set xs \<Longrightarrow> no_trace m') \<Longrightarrow> no_trace (mapME_x f xs)"
+  unfolding mapME_x_def
+  by (fastforce intro: no_trace_sequenceE_x)
+
+lemma no_trace_filterM[no_trace_cond]:
+  "(\<And>m. m \<in> set ms \<Longrightarrow> no_trace (P m)) \<Longrightarrow> no_trace (filterM P ms)"
+  by (induct ms; wpsimp wp: no_trace_terminal split_del: if_split)
+
+lemma no_trace_zipWithM_x[no_trace_cond]:
+  "(\<And>x y. no_trace (f x y)) \<Longrightarrow> no_trace (zipWithM_x f xs ys)"
+  unfolding zipWithM_x_def zipWith_def
+  by (fastforce intro: no_trace_sequence_x)
+
+lemma no_trace_zipWithM[no_trace_cond]:
+  "(\<And>x y. no_trace (f x y)) \<Longrightarrow> no_trace (zipWithM f xs ys)"
+  unfolding zipWithM_def zipWith_def
+  by (fastforce intro: no_trace_sequence)
+
+lemma no_trace_foldM[no_trace_cond]:
+  "(\<And>x y. no_trace (m x y)) \<Longrightarrow> no_trace (foldM m xs a)"
+  unfolding foldM_def
+  by (induct xs; wpsimp wp: no_trace_terminal)
+
+lemma no_trace_foldME[no_trace_cond]:
+  "(\<And>x y. no_trace (m x y)) \<Longrightarrow> no_trace (foldME m a xs)"
+  unfolding foldME_def
+  by (induct xs; wpsimp wp: no_trace_terminal)
+
+lemma no_trace_maybeM[no_trace_cond]:
+  "\<forall>x. no_trace (f x) \<Longrightarrow> no_trace (maybeM f t)"
+  unfolding maybeM_def
+  by (fastforce intro: no_trace_terminal split: option.splits)
+
+lemma no_trace_notM[no_trace_cond]:
+  "no_trace A \<Longrightarrow> no_trace (notM A)"
+  unfolding notM_def
+  by (wpsimp wp: no_trace_terminal)
+
+lemma no_trace_ifM[no_trace_cond]:
+  "\<lbrakk> no_trace P; no_trace a; no_trace b \<rbrakk> \<Longrightarrow> no_trace (ifM P a b)"
+  unfolding ifM_def by wpsimp
+
+lemma no_trace_ifME[no_trace_cond]:
+  "\<lbrakk> no_trace P; no_trace a; no_trace b \<rbrakk> \<Longrightarrow> no_trace (ifME P a b)"
+  unfolding ifME_def by wpsimp
+
+lemma no_trace_whenM[no_trace_cond]:
+  "\<lbrakk> no_trace P; no_trace f \<rbrakk> \<Longrightarrow> no_trace (whenM P f)"
+  unfolding whenM_def
+  by (wpsimp wp: no_trace_terminal no_trace_cond)
+
+lemma no_trace_orM[no_trace_cond]:
+  "\<lbrakk> no_trace A; no_trace B \<rbrakk> \<Longrightarrow> no_trace (orM A B)"
+  unfolding orM_def
+  by (wpsimp wp: no_trace_terminal no_trace_cond)
+
+lemma no_trace_andM[no_trace_cond]:
+  "\<lbrakk> no_trace A; no_trace B \<rbrakk> \<Longrightarrow> no_trace (andM A B)"
+  unfolding andM_def
+  by (wpsimp wp: no_trace_terminal no_trace_cond)
+
+\<comment> \<open>FIXME: This is true but is it useful to even state?
+          The parallel composition of traceless functions doesn't make much sense\<close>
+lemma no_trace_parallel[no_trace_cond]:
+  "\<lbrakk> no_trace f; no_trace g \<rbrakk> \<Longrightarrow> no_trace (parallel f g)"
```

Comment:
I think it's useful to have -- you want trivial goals to go away automatically, and this will enable that. It's unlikely to occur, but I could imagine it popping up in some kind of contradiction or things like that.

Reply:
Good point. I'll remove the FIXME then but might leave a comment since I imagine we might otherwise go through the same conversation at some point in the future..